tpu-client/src/nonblocking/tpu_client.rs

```diff
@@ -503,17 +503,55 @@ where
         Self::new_with_connection_cache(rpc_client, websocket_url, config, connection_cache).await
     }
 
+    /// Try to create LeaderTpuService
+    /// Retries until successful or timeout is reached
+    async fn try_create_leader_tpu_service(
+        rpc_client: Arc<RpcClient>,
+        websocket_url: &str,
+        exit: Arc<AtomicBool>,
+        timeout_seconds: u64,
+    ) -> Result<LeaderTpuService> {
+        let start = tokio::time::Instant::now();
+        loop {
+            if start.elapsed().as_secs() > timeout_seconds {
+                return Err(TpuSenderError::Custom(format!(
+                    "Failed to create LeaderTpuService within {timeout_seconds}s timeout"
+                )));
+            }
+
+            match LeaderTpuService::new(
+                rpc_client.clone(),
+                websocket_url,
+                M::PROTOCOL,
+                exit.clone(),
+            )
+            .await
+            {
+                Ok(service) => return Ok(service),
+                Err(_) => {
+                    warn!("Failed to create TpuLeaderService. Will retry in 1 second");
+                    sleep(Duration::from_secs(1)).await;
```

Comment:
If this is a common situation, should we really warn on the first failure?
"Warnings that are fine and should be ignored" is a suboptimal practice, as it erodes the trust in what the system is saying about itself.

In this particular case, the expectation is that the client will be constructed in `timeout_seconds`.  And we error out if that does not happen.
Do we even want to report anything here?

Reply:
ahh ok fair enough. the caller of this function should know it may take `timeout_seconds`, so makes sense not to warn. will update